rasa/core/policies/ensemble.py

```diff
@@ -737,6 +707,64 @@ def _check_policy_for_forms_available(
         )
 
 
+def _training_events_from_trackers(
+    training_trackers: List[DialogueStateTracker],
+) -> Dict[Text, Set[Event]]:
+    """Creates a dictionary of action names and events that follow these actions.
+
+    Args:
+        training_trackers: the trackers used for training
+
+    Returns:
+        a dictionary of action names and events that follow these actions
+    """
+    events_metadata = defaultdict(set)
+
+    for t in training_trackers:
+        tracker = t.init_copy()
+        for event in t.events:
+            tracker.update(event)
+            if isinstance(event, ActionExecuted):
+                continue
+
+            action_name = tracker.latest_action_name
+            if action_name:
+                events_metadata[action_name].add(event)
+
+    return events_metadata
+
+
+def create_action_fingerprints(
+    trackers: List[DialogueStateTracker],
+) -> Dict[Text, Dict[Text, List[Text]]]:
+    """Fingerprint each action using the events it created during train.
+
+    This allows us to emit warnings when the model is used
+    if an action does things it hasn't done during training,
+    or if rules are incomplete.
+
+    Args:
+        trackers: the list of trackers
+
+    Returns:
+        a nested dictionary of action names and slots and active loops
+            that this action sets
+    """
+    training_events = _training_events_from_trackers(trackers)
+    if not training_events:
+        return {}
+
+    action_fingerprints = {}
+    for k, vs in training_events.items():
+        slots = list({v.key for v in vs if isinstance(v, SlotSet)})
+        active_loops = list({v.name for v in vs if isinstance(v, ActiveLoop)})
+        action_fingerprints[k] = {
+            ACTION_FINGERPRINT_SLOTS: slots,
+            ACTION_FINGERPRINT_ACTIVE_LOOPS: active_loops,
+        }
+    return action_fingerprints
```

Comment:
Let's use a typed class instead of a dict.

Reply:
then we also don't need more constants

Reply:
this is an old dict, I'm not sure where it is used and how it is persisted.

Reply:
I think yes as we can skip all these constants then.

Reply:
I reused `SLOTS` and `ACTIVE_LOOP` constants because they actually have to be coupled, so I removed these constants anyhow

Reply:
I suggest let's not over engineer it and keep a dict, it looks accurate right now

Reply:
also, it could happen that action doesn't set any slots or loops, so these keys are not guaranteed to be in there